Source/Core/Common/Image.cpp

```diff
@@ -0,0 +1,183 @@
+// Copyright 2016 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#include <Common/Image.h>
+
+#include <csetjmp>
+#include <cstdlib>
+#include <cstring>
+#include <memory>
+#include <png.h>
+#include <vector>
+
+#include "Common/CommonTypes.h"
+#include "Common/Logging/Log.h"
+
+namespace Common
+{
+static int MultiplyAlpha(int alpha, int color)
+{
+  const int temp = (alpha * color) + 0x80;
+  return ((temp + (temp >> 8)) >> 8);
+}
+
+static void PremultiplyData(png_structp png, png_row_infop row_info, png_bytep data)
+{
+  for (png_size_t i = 0; i < row_info->rowbytes; i += 4, data += 4)
+  {
+    const png_byte alpha = data[3];
+    u32 w;
+
+    if (alpha == 0)
+    {
+      w = 0;
+    }
+    else
+    {
+      png_byte red = data[0];
+      png_byte green = data[1];
+      png_byte blue = data[2];
+
+      if (alpha != 0xff)
+      {
+        red = MultiplyAlpha(alpha, red);
+        green = MultiplyAlpha(alpha, green);
+        blue = MultiplyAlpha(alpha, blue);
+      }
+      w = (alpha << 24) | (red << 16) | (green << 8) | (blue << 0);
+    }
+
+    std::memcpy(data, &w, sizeof(w));
+  }
+}
+
+struct ReadProgress
+{
+  const u8* current;
+  const u8* end;
+};
+
+static void ReadCallback(png_structp png, png_bytep data, png_size_t size)
+{
+  ReadProgress* progress = static_cast<ReadProgress*>(png_get_io_ptr(png));
+  for (size_t i = 0; i < size; ++i)
+  {
+    if (progress->current >= progress->end)
+      png_error(png, "Read beyond end of file");
```

Comment:
Missing return here?

Reply:
The error handler uses longjmp, so there's no need for a return here.

Reply:
That's not something you expect when reading this code. Specially since `png_error` isn't documented to be no-return. I had to look in the libpng source code after you pointed it out.

Reply:
It is documented to be no-return in the libpng manual:

> Error handling in libpng is done through png_error() and png_warning(). Errors handled through png_error() are fatal, meaning that png_error() should never return to its caller. Currently, this is handled via setjmp() and longjmp() (unless you have compiled libpng with PNG_SETJMP_NOT_SUPPORTED, in which case it is handled via PNG_ABORT()), but you could change this to do things like exit() if you should wish.

But that's a relatively long document to read, so I'll add a comment to this code about it.